File: DataStructure/Deque/q402-remove-k-digits.py
# 给定一个以字符串表示的非负整数 num，移除这个数中的 k 位数字，使得剩下的数字最小。
#
#  注意:
#
#
#  num 的长度小于 10002 且 ≥ k。
#  num 不会包含任何前导零。
#
#
#  示例 1 :
#
#
# 输入: num = "1432219", k = 3
# 输出: "1219"
# 解释: 移除掉三个数字 4, 3, 和 2 形成一个新的最小的数字 1219。
#
#
#  示例 2 :
#
#
# 输入: num = "10200", k = 1
# 输出: "200"
# 解释: 移掉首位的 1 剩下的数字为 200. 注意输出不能有任何前导零。
#
#
#  示例 3 :
#
#
# 输入: num = "10", k = 2
# 输出: "0"
# 解释: 从原数字移除所有的数字，剩余为空就是0。
#
#  Related Topics 栈 贪心算法
#  👍 376 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack, remain = [], len(num) - k
        for digit in num:
            while k and stack and stack[-1] > digit:
                stack.pop()
                k -= 1
            stack.append(digit)
        return ''.join(stack[:remain]).lstrip('0') or '0'
# leetcode submit region end(Prohibit modification and deletion)

File: DataStructure/Deque/q402_remove_k_digits.java
package DataStructure.Deque;

import java.util.*;

public class q402_remove_k_digits {
    // 单调栈
    public String removeKdigits(String num, int k) {
        int n = num.length();
        int remain = n - k;
        if (remain == 0) return "0";

        Deque<Integer> stack = new ArrayDeque<>();
        for (char ch: num.toCharArray()) {
            int curNum = Integer.valueOf(ch + "");
            while (k > 0 && !stack.isEmpty() && stack.peek() > curNum) {
                // 新来的数更小，说明可以把栈顶pop，放入更小的数字 ->> 维护单调递增栈
                stack.pop();
                k--;
            }
            stack.push(curNum);
        }
        while (!stack.isEmpty() && k-- > 0) stack.pop(); // ?勿漏！法1) 还需pop，直至k为0!!!!
        while (!stack.isEmpty() && stack.peekLast() == 0) stack.removeLast(); //法1） ?队头/栈底=Last：前导0去除, 不可pop！
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.insert(0, stack.pop());
        }
        // 法2）先：若本就递增，则stack无法pop，需在此截取前remain个↓ (避免先lstrip导致截取到后面的大数)
        // if (sb.length() > remain) sb = new StringBuilder(sb.substring(0, remain));
        // 法2）后：lstrip('0')
        // while (sb.length() > 0 && sb.charAt(0) == '0') sb.deleteCharAt(0);
        return sb.length() == 0? "0" : sb.toString();
    }
}
